feat(combinator): add separated

vwkd opened the pull request:
This attempts to add a `separated` parser that accepts a range similar to `repeat`.

Towards #98.

I tried to copy as much logic as possible from existing code. I'm unfamiliar with the code and may have missed some things, e.g. checkpoint resets, infinite loop checks, or correct error types. Please double check carefully and feel free to edit if it can be improved.

epage commented on src/combinator/multi.rs:
```diff
@@ -238,17 +238,8 @@ where
-        let len = input.eof_offset();
         match parse.parse_next(input) {
             Ok(value) => {
-                // infinite loop check: the parser must always consume
```
Can you split this out into an issue to discuss?

My biggest concern for why I want to talk through this more is the inconsistent behavior for `repeat`, depending on what range is used

vwkd replied:
Agree, split out into #366 with #365 to discuss it first.

epage commented on src/combinator/multi.rs:
```diff
@@ -333,166 +333,367 @@ where
 #[doc(alias = "sep_by")]
+#[doc(alias = "sep_by1")]
 #[doc(alias = "separated_list0")]
-pub fn separated0<I, O, C, O2, E, P, S>(mut parser: P, mut sep: S) -> impl Parser<I, C, E>
```
Can you keep `separated0` and just make it deprecated?

How I normally handle these transitions is I add the new feature in one commit, make the next commit update callers to it, and then the following commit deprecates it.  Keeps the commits small and buildable.

vwkd replied:
Done.

epage commented on src/combinator/multi.rs:
```diff
@@ -333,166 +333,367 @@ where
+    S: Parser<I, O2, E>,
+    E: ParserError<I>,
+{
+    let mut acc = C::initial(None);
```
To make reviews easier, I'd recommend renaming variables to a follow up commit

vwkd replied:
The naming is consistent with the newer `repeat` functions. Since the existing `separated0` and `separated1` are now kept, the diff shouldn't be impacted anymore.

epage replied:
Ideally the `separated0` and `separated1` would call into the new methods and the code would be arranged so the diffs would show that no behavior was changed in the bodies.  This also helps with doing `git blame`.  Like I said, later commits can do cleanups (much like your refactor PR for moving the functions, that was a byproduct of me doing the same thing and not getting to cleaning it up)

epage commented on src/combinator/multi.rs:
```diff
@@ -497,18 +497,13 @@ where
                     ));
                 }
 
-                let len = i.eof_offset();
```
As this is removing extra work done compared to the original implementation, I'd recommend just squashing this commit into the prior commit

vwkd replied:
Yes, I've added them back and moved the removal to #366.

epage commented on src/combinator/multi.rs:
```diff
@@ -333,166 +333,367 @@ where
+    let start = input.checkpoint();
+    match parser.parse_next(input) {
+        Err(ErrMode::Backtrack(e)) => {
+            if min > 0 {
+                return Err(ErrMode::Backtrack(e.append(input, ErrorKind::Many)));
+            } else {
+                input.reset(start);
+                return Ok(acc);
+            }
```
nit: swapping the cases so the condition is `min == 0` I feel would be cleaner / clearer

vwkd replied:
True. Fixed now.

epage commented on src/combinator/multi.rs:
```diff
@@ -333,166 +333,367 @@ where
+    Ok(acc)
+}
+
+fn separated_m_n_<I, O, C, O2, E, P, S>(
```
Can you add the infinite checks?

vwkd replied:
Added. Reversal commit in #366.

epage commented:
Thanks for moving this forward!

epage commented on src/combinator/multi.rs:
```diff
@@ -504,6 +504,375 @@ where
+#[doc(alias = "separated_list1")]
+#[doc(alias = "separated_m_n")]
+#[inline(always)]
+pub fn separated<I, O, C, O2, E, P, S>(
```
This commit should also includes tests for the new modes, matching the test cases covered for `repeat`

vwkd replied:
Done.

vwkd commented on src/combinator/multi.rs:
```diff
@@ -333,166 +333,367 @@ where
-#[doc(alias = "sep_by1")]
-#[doc(alias = "separated_list1")]
-pub fn separated1<I, O, C, O2, E, P, S>(mut parser: P, mut sep: S) -> impl Parser<I, C, E>
+fn separated_n_<I, O, C, O2, E, P, S>(
```
Added. Reversal commit in #366.

vwkd commented:
This should be good to go for another round.

The infinite loop checks for the element parser (not the separator parser) in `separated0_` currently make one test fail. Not sure what to do here.

epage commented:
> Hope this is now better and easier to review. Just to stress again, I'm not sure how correct this code is. Please review carefully.
> 
> Currently, the infinite loop checks for the element parser (not the separator parser) in `separated0_` make one test fail.

This PR should not be changing behavior and duplicating the `separated0` and `separated1` bodies, its hard to tell what behavior changed for us to talk about it.

vwkd commented:
I must have misunderstood the request to keep `separated0` and deprecating it.

As requested, I've now changed it back such that `seperated0` becomes `separated0_` (and similar for `1`), added tests, and moved the variable renames into a separate commit. The diff should be readable now. I've also removed again the infinite loop check for the element parser (not the separator parser), which was not in the existing implementation, which now makes all tests pass.

Please note, I kept the implementation and the updating call sites in separate commits as requested, but since the signature of the functions change, it can't be used at the call sites, which means the implementation commit (`feat(combinator): add separated`) won't compile.

Slightly off topic: Maybe it would make sense to adopt a squash/rebase flow instead of a merge flow, such that individual commits of the PR can be purely for ease of readability and review, since they don't need to compile as the whole PR gets squashed into a single commit on the main branch. A linear history and readable commit log is a nice bonus.

epage commented:
> I must have misunderstood the request to keep `separated0` and deprecating it.
> 
> As requested, I've changed it back such that `seperated0` becomes `separated0_` (and similar for `1`). I've added tests, and moved the variable renames into a separate commit, such that the diff is more readable[1](#user-content-fn-1-6221bc56f380ac0f7f3903667bebb59e). I've again removed the extra infinite loop checks for the element parsers (not the existing one for the separator parsers) - which was not in the existing implementation - such that all tests pass.

This change should not be a breaking change.  You can both split out `separated0_` for reuse **and** keep `separator0`.  See #241.
 
> Slightly off topic: Maybe it would make sense to adopt a squash/rebase flow instead of a merge flow, such that individual commits of the PR can be purely for ease of readability and review, since they don't need to compile as the whole PR gets 

Squash/rebase workflows lose valuable history and intentionally sending someone broken commits to review can be confusing.

vwkd commented:
I think I understand the request now. I've kept deprecated `separated0` that reuse `separated0_` internally (similar for `1`) and a separated commit to remove them. This should keep each commit buildable.

epage commented on src/combinator/multi.rs:
```diff
@@ -468,51 +468,6 @@ where
-#[doc(alias = "sep_by")]
-#[doc(alias = "separated_list0")]
-#[deprecated]
-pub fn separated0<I, O, C, O2, E, P, S>(mut parser: P, mut sep: S) -> impl Parser<I, C, E>
```
One of my pieces of guidance for this PR is that there should be no breaking changes.  Please drop this commit from the PR.

vwkd replied:
Done.

epage commented on src/combinator/multi.rs:
```diff
@@ -463,45 +611,218 @@ where
+                        return Err(e);
+                    }
+                    Ok(o) => {
+                        // infinite loop check
+                        if input.eof_offset() == len {
```
Infinite loop checks only go on the separator

vwkd replied:
Removed.

epage commented on src/combinator/multi.rs:
````diff
@@ -454,6 +601,7 @@ where
 /// ```
 #[doc(alias = "sep_by1")]
 #[doc(alias = "separated_list1")]
+#[deprecated]
````
Don't forget to list the `since` and `note`

vwkd replied:
Could you specify what values you desire?

vwkd replied:
Done. Assumed the next version will be `0.5.19`.

epage commented:
Everything should be good to go after these last couple of comments are resolved!  Thanks for your work on this!

vwkd commented:
Addressed the remaining comments.

To lessen the previously breaking change, before I had considered proposing this in a follow up. But since this is now a non-breaking additive change, we could take the opportunity to choose the right name now which would also avoid another breaking change in the future. I think naming it `separate` (verb, instead of adjective `separated`) fits better and also aligns with the existing `repeat`.

epage commented:
If you'd like to discuss name changes, I'd recommend looking into whats been done before on https://github.com/winnow-rs/winnow/discussions/95 and posting your idea there.

epage commented:
Thanks!
